Source/VCGeneration/Counterexample.cs

```diff
@@ -667,4 +674,53 @@ public virtual void OnVCResult(VCResult result)
       Contract.Requires(result != null);
     }
   }
+  
+  // Intercepts callbacks to record counter-examples in splits
+  class SplitVerifierCallback : VerifierCallback {
+    private readonly VerifierCallback underlying;
+    private readonly Split split;
+
+    public SplitVerifierCallback(VerifierCallback underlying, Split split) : base(CoreOptions.ProverWarnings.None) {
+      this.underlying = underlying;
+      this.split = split;
+    }
+    // reason == null means this is genuine counterexample returned by the prover
+    // other reason means it's time out/memory out/crash
+    public override void OnCounterexample(Counterexample ce, string /*?*/ reason)
+    {
+      split.Counterexamples.Add(ce);
```

Comment:
Why don't you do place `split.Counterexamples.Add(ce)` before the call to `OnCounterexample` in `SplitAndVerifyWorker`, then you don't need this class.

Reply:
You mean, before this call?
`callback.OnCounterexample(split.ToCounterexample(split.Checker.TheoremProver.Context), msg);`
The problem is, `OnCounterExample` is not called only from `SplitAndVerifyWorker`. It's also called from `VCGen` 2 times. Look for `ErrorReporter.OnModel`, it uses the callback defined in `Split.cs` line `reporter = new VCGen.ErrorReporter(...`, which callback is a `SplitVerifyCallback` that records the counter-examples in the split as you have seen.